src/xinteractiveshell.hpp

```diff
@@ -0,0 +1,73 @@
+#include "pybind11/pybind11.h"
+#include "xdisplay.hpp"
+
+#ifdef __GNUC__
+    #pragma GCC diagnostic ignored "-Wattributes"
+#endif
+
+namespace py = pybind11;
+using namespace pybind11::literals;
+
+namespace xpyt 
+{
+
+    struct hooks_object
+    {
+        hooks_object() {}
+        static inline void show_in_pager(py::str data, py::kwargs)
+        {
+            xdisplay(py::dict("text/plain"_a=data), {}, {}, py::dict(), py::none(), py::none(), false, true);
+        }
+
+    };
+
+    class XInteractiveShell {
+
+    private:
+        py::module _ipy_process;
+        py::module _magic_core;
+        py::module _magics_module;
+        py::module _extension_module;
+
+    public:
+        XInteractiveShell();
+
+        //required by cd magic and others
+        py::dict db;
+        py::dict user_ns;
+
+        //required by extension_manager
+        py::object builtin_trap;
+        py::str ipython_dir;
+
+        //required by pushd
+        py::list dir_stack;
+        py::str home_dir;
+
+        py::object magics_manager;
+        py::object extension_manager;
+
+        hooks_object hooks;
```

Comment:
Do we need write access for all of these properties? If not, they should be moved to the private section and getters should be provided. You can bind then the getters with `def_property_readonly`.